stdlib/mlang/language-composer.mc

```diff
@@ -0,0 +1,355 @@
+  sem handleDecl langStr includes ctx = 
+  | decl & DeclSem d ->
+    let identStr = nameGetStr d.ident in 
+    let findMatchingInfo : String -> Option DeclInfo = lam incl.
+      mapLookup (incl, identStr) ctx.langMap in 
+    let foundIncludes : [DeclInfo] = mapOption findMatchingInfo includes in 
+    
+    let conflicts = filter (lam i. or (isTypeInfo i) (isSynInfo i)) foundIncludes in 
+    let errors = cons (d.info, nameGetStr d.ident) (map extractInfoName conflicts) in 
+
+    if not (null conflicts) then
+      errorMulti errors "The declared sem has an identifier that conflicts with included types!"
+    else
+      let includedSems = filter isSemInfo foundIncludes in 
+
+      let includes = map projIdent includedSems in 
+      (ctxWithDeclInfo ctx (langStr, nameGetStr d.ident) (decl2info langStr decl), 
+       DeclSem {d with includes = includes})
+  | decl & DeclSyn d ->
+    let identStr = nameGetStr d.ident in 
+    let findMatchingInfo : String -> Option DeclInfo = lam incl.
+      mapLookup (incl, identStr) ctx.langMap in 
+    let foundIncludes : [DeclInfo] = mapOption findMatchingInfo includes in 
+    
+    let conflicts = filter isTypeInfo foundIncludes in 
+
+    let errors = cons (d.info, nameGetStr d.ident) (map extractInfoName conflicts) in 
+
+    if not (null conflicts) then
+      errorMulti errors "The declared syn has an identifier that conflicts with included types!"
+    else
+      let includedSyns = filter isSynInfo foundIncludes in 
+
+      let includes = map projIdent includedSyns in 
+      let info = {ident = d.ident, info = d.info} in 
+      (ctxWithDeclInfo ctx (langStr, nameGetStr d.ident) (decl2info langStr decl), 
+       DeclSyn {d with includes = includes})
+  | decl & DeclType d -> 
+    let identStr = nameGetStr d.ident in 
+    let findMatchingInfo : String -> Option DeclInfo = lam incl.
+      mapLookup (incl, identStr) ctx.langMap in 
+    let conflicts : [DeclInfo] = mapOption findMatchingInfo includes in 
+
+    let errors = cons (d.info, nameGetStr d.ident) (map extractInfoName conflicts) in 
+
+    if not (null conflicts) then
+      errorMulti errors "The declared type has an identifier that conflicts with included types or syns!"
+    else
+      let info = decl2info langStr decl in 
+      (ctxWithDeclInfo ctx (langStr, nameGetStr d.ident) info, decl)
+  | _ -> error "Only Type, Syn, and Sem declarations can be contained inside of a langauge!"
+
+  sem addImplicitIncludes langStr includes definedSynsSems =
+  | ctx ->
+    -- We are going to remove elements from the ctx.langMap that
+    -- (1) do not belong to an included langauge
+    -- (2) that have already been included explicitly through a syn or sem
+    -- (3) that are not Type declarations.
+    let pred = lam k. lam v. 
+      match k with (origLang, ident) in 
+        (and (not (isTypeInfo v))
+             (and (seqMem eqString includes origLang)
```

Comment:
`seqMem` smells like a performance bug to me

Reply:
How so? A set would typically be more efficient, but it is quite common to have to check for membership in a list anyways right? It doesn't really make sense to me to construct a set each time.

Reply:
`pred` will be called multiple times (so at least complexity "size of ctx.langMap" * "size of includes", vs. "size of ctx.langMap" * "log of size of includes" + "n log n of size of includes").

If you make `includes` a set/map you could also work with `mapIntersect`.